### PR TITLE
Fix ``super-init-not-called`` if parent or ``self`` is a ``Protocol``

### LINKED ISSUE #4790: false positive super-init-not-called for subclasses of typing.Protocol
### Bug description

Given the example code below, both `MyProtocol.__init__()` and `ExplicitProtocolImplimentation.__init__()` are being flagged with `super-init-not-called`.

```python
from abc import abstractmethod
from typing import Any, Protocol


class MyProtocol(Protocol):
    @abstractmethod
    def __init__(self, foo: Any) -> None:
        raise NotImplementedError


class ExplicitProtocolImplimentation(MyProtocol):
    def __init__(self, foo: Any) -> None:
        print(foo)
```
The results are the same with/without using `@abstractmethod`
and/or replacing `typing.Protocol` with `typing_extensions.Protocol`.

### Configuration

_No response_

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
test.py:7:4: W0231: __init__ method from base class 'Protocol' is not called (super-init-not-called)
test.py:12:4: W0231: __init__ method from base class 'MyProtocol' is not called (super-init-not-called)
```

### Expected behavior

I would expect that neither of these would be flagged with `super-init-not-called`.

While calling `super().__init__()` here won't cause a runtime error, it also won't do anything. As noted in [PEP 544](https://www.python.org/dev/peps/pep-0544/), _[Protocols cannot be instantiated](https://www.python.org/dev/peps/pep-0544/#subtyping-relationships-with-other-types)_. This is enforced by the removal of the implementation of `MyProtocol.__init__()` at runtime.

I would also expect that trying to use `super().__init__()` in `MyProtocol.__init__()` or `ExplicitProtocolImplimentation.__init__()` would result in a warning since we can deduce that it is unnecessary (e.g. `superfluous-super`) but this could be addressed in a separate issue as a feature request.

### Pylint version

```shell
pylint 2.9.6
astroid 2.6.5
Python 3.9.6 (default, Jul 15 2021, 14:52:44) 
[GCC 9.3.0]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #4790.

Added typing and `confidence` as well.

### CODE REVIEW

File: tests/functional/s/super/super_init_not_called.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+super-init-not-called:50:4:51:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE
```
Comment (Pierre-Sassoulas) [inline tests/functional/s/super/super_init_not_called.txt:1]:
```suggestion
super-init-not-called:51:4:52:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Nice, thank you @DanielNoord !

File: pylint/checkers/classes/class_checker.py [hunk 2]
```diff
@@ -1086,12 +1087,12 @@ def visit_functiondef(self, node: nodes.FunctionDef) -> None:
         self._check_useless_super_delegation(node)
         self._check_property_with_parameters(node)
 
-        klass = node.parent.frame(future=True)
+        klass = node.parent.frame(future=True)  # type: nodes.ClassDef # See is_method()
```
[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylint/checkers/classes/class_checker.py:1090]: The check on L1084 makes sure that this is true. However, `mypy` doesn't pick up on that.

[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:1090]: Ha alright, ``is_method`` was out of side in the review.
```suggestion
        # 'is_method()' is called and makes sure that this is a 'nodes.ClassDef'
        klass = node.parent.frame(future=True)
```
Maybe clearer ?

**Review by Pierre-Sassoulas — APPROVED** [review]
Great fix again 👌

**Comment by Pierre-Sassoulas** [comment]
Look like there is an issue in the functional test when inference is not Undefined ?

**Comment by DanielNoord** [comment]
Oh no, stupid mistake. Since I updated the `confidence` for this message all tests that check for this message should be updated. This is a test in another file that I don't change but for which the output should be updated. Fix coming in 30 seconds.

**Comment by DanielNoord** [comment]
Hmm some version and OS dependent stuff it seems like. Going to take another look tonight!

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Do we know a contributor who works on Windows? I have no idea why this doesn't work on `3.10` on Windows. Is `Protocol` missing in `typing_extensions` on that version perhaps?

**Comment by Pierre-Sassoulas** [comment]
> Do we know a contributor who works on Windows?

It's like searching for teeth in a hen's mouth 😄 I'm going to see if I can set up a VM some day. I will need a license before being able to do that anyway.

>  Is Protocol missing in typing_extensions on that version perhaps

The ``6: import-error`` makes me think that you're right.

**Comment by DanielNoord** [comment]
> > Do we know a contributor who works on Windows?
> 
> It's like searching for teeth in a hen's mouth 😄 I'm going to see if I can set up a VM some day. I will need a license before being able to do that anyway.

I was thinking the same about Linux. I really need it for that `distutils` fix...

> > Is Protocol missing in typing_extensions on that version perhaps
> 
> The `6: import-error` makes me think that you're right.

But I don't understand why. I didn't see any issues about it in the `typing` repo... 😅

**Comment by DanielNoord** [comment]
Tests are fixed no!

**Review by Pierre-Sassoulas — APPROVED** [review]
👍

**Pull request merged**